### ISSUE TITLE
[Due for payment 2026-04-16] Migrate the main chat list to FlashList

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#33725]
HOLD on:

- Comment linking to be 100% complete
- The new architecture to be enabled (this is not a hard requirement but I have a hunch that this should wait on that too)

### Problem
The main chat list is one of the most, if not _the most_ important component in our app. If you scroll far or fast on this list, you may see frames drop. Furthermore, there are some known performance issues with this list ([example](https://github.com/Expensify/App/issues/30495)). We have already migrated almost every other virtualized list in our app from FlatList to FlashList, because its performance is much better.

### Solution
Let's build support for bidirectional pagination in FlashList, then enable it on the main chat list in E/App.

**Comment by roryabraham** [comment]
I expect this to be on HOLD until sometime in February, if not later, so I'm moving this back to monthly

**Comment by roryabraham** [comment]
cc @muttmuure adding this to the New Expensify performance improvements project. I acknowledge that the problem statement is a bit of a reverse-solution but I think it's a good bet that this will provide substantial performance improvements, as FlashList is just much better for performance than FlatList. We've established that elsewhere where we've migrated to FlashList

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by roryabraham** [comment]
on HOLD for comment linking

**Comment by garrettmknight** [comment]
@roryabraham we good to take this off hold? If so, where do you think it might fit in a wave, if any? I'm not 100% on the significance of this migration beyond general, better performance.

**Comment by roryabraham** [comment]
Good question. As far as I know there's no significance beyond general, better performance of chat screens.

**Comment by perunt** [comment]
@roryabraham, you’re absolutely right about the performance perks we’d get with FlashList—better speed, faster rendering, the whole deal. But there's a bit of a twist when it comes to the bidirectional scrolling we need. Right now, adding that feature to FlashList can lead to some choppy experiences. It might slow us down a bit to get that smoothed out.

On another note, Irisjae just released some interesting updates (check it out [here](https://github.com/irisjae/recyclerlistview/blob/master/README.md)). He’s added something called `preserveVisiblePosition`, which is aimed directly at fixing those choppy scrolling issues. I haven't tried it yet, but it could work for us.

**Comment by perunt** [comment]
@roryabraham, you’re absolutely right about the performance perks we’d get with FlashList—better speed, faster rendering, the whole deal. But there's a bit of a twist when it comes to the bidirectional scrolling we need. Right now, adding that feature to FlashList can lead to some choppy experiences. It might slow us down a bit to get that smoothed out.

We have PR with the implementation of a bidirectional list, but it is a bit stale (https://github.com/Shopify/flash-list/pull/824#issuecomment-2100450355)
also would be nice to double-check this approach (https://github.com/irisjae/recyclerlistview/blob/master/README.md). This guy added something called preserveVisiblePosition, which aims to fix those choppy scrolling issues. I have yet to try it, but it could work for us.

**Comment by garrettmknight** [comment]
Still on hold.

**Comment by muttmuure** [comment]
I think this can come off hold now?

**Comment by garrettmknight** [comment]
@roryabraham can this come of hold?

**Comment by muttmuure** [comment]
@janicduplessis is going to investigate this

**Comment by janicduplessis** [comment]
I will take this :)

**Comment by janicduplessis** [comment]
After first investigation of this there are 2 main things we need to implement:

1. Support `onStartReached`. This one is pretty simple, I already have a working prototype.
2. Support `maintainVisibleContentPosition` (or some alternative implementation that does the same thing). I found why `maintainVisibleContentPosition` does not work with FlashList, it has to do with the view hierarchy it creates. The `maintainVisibleContentPosition` implementation assumes `ScrollView -> ContentView -> Cells`, but FlashList creates `ScrollView -> ContentView -> AutoLayout -> Cells`.

For `#2` I am currently investigating the different approaches we can use to fix this.

**Comment by roryabraham** [comment]
I'm headed on parental leave. @mountiny going to reassign to you as this is part of #newdot-quality.

**Comment by mountiny** [comment]
@janicduplessis How is this one looking? What are the next steps and ETA?

**Comment by janicduplessis** [comment]
Picking up work on this this week. Will provide another update and ETA soon.

**Comment by janicduplessis** [comment]
Update: I abandoned the maintainVisibleContentPosition route as it isn't really compatible with how recyclerlistview handles virtualization.

Instead I am currently looking at using a redesigned version of flash-list / recyclerlistview, the changes made are described in details [here](https://github.com/irisjae/recyclerlistview/blob/master/README.md) (really suggest going through it, very interesting). In summary it changes the virtualization algorithm to have first class support for keeping the visible content position.

I managed to get it working in expensify app, the main work needed was to get it working on new arch and add support for onStartReached. It seem to work really well except for a pretty big flicker when it adjusts the position of items in one case. I have some ideas on how to fix this. I am pretty optimistic that this is a great solution and will continue working in that direction.

Will post some demo and code later.

**Comment by muttmuure** [comment]
@janicduplessis how is this going?

**Comment by janicduplessis** [comment]
I've been working on some other tasks, but should be good to come back on this soon.

**Comment by garrettmknight** [comment]
@janicduplessis @hannojg same question! any updates here?

**Comment by garrettmknight** [comment]
@mountiny any thoughts on this one?

**Comment by mountiny** [comment]
@hannojg @janicduplessis can you please post another update? thanks!

**Comment by hannojg** [comment]
We haven't progressed with that task yet. Janic unfortunately become unavailable for working on this.
I am currently trying to reschedule this task in our team!
Will keep you posted!

**Comment by mallenexpensify** [comment]
Posted in #expert-contributors
> We're looking for someone to take this on. 
[@vit](https://expensify.slack.com/team/U0228UGAF3R) [@Hanno Gödecke](https://expensify.slack.com/team/U043ZHPLVQQ)
, any ideas who might be the best fit?

**Comment by hannojg** [comment]
We re-evaluated this ticket internally and think that it's probably not worth the effort.
`FlashList` isn't really build with bi-directional scrolling in mind and forcing it to do it will only gain mediocre result that wouldn't justify the high amount of time we'd need to invest.

We still want to work on making the chat list as fast as it can be. We want to build a few PoC (where we keep bidirectional scrolling in mind) and evaluate those. This will probably be a mix of:

- Write our own implementation
- Legend list 
- Other solutions we might come across

We might want to either:
- close this ticket and open a new one
- convert this ticket to be a general list performance investigation ticket

**Comment by mallenexpensify** [comment]
Thanks for the update @hannojg , I put the issue on hold for now.  There's no big rush, I'm just going through HIGH and CRITICAL issues on the #quality project to ensure we're properly aligned and prioritized.

**Comment by garrettmknight** [comment]
Still on hold.

**Comment by garrettmknight** [comment]
Still on hold.

**Comment by garrettmknight** [comment]
@mallenexpensify We should probably address this separately as part of a specific project. I'm not sure this open GH is doing much for us now that the solution isn't as straightforward. What do you think?

**Comment by mountiny** [comment]
I agree, I think this one has stalled as the solutions rendered as not so clear. We might be good closing this and if there will be some proposals for using different List that are promising - go ahead with that

**Comment by chrispader** [comment]
Apparently, the issues that were blocking us from migrating to FlashList have been fixed recently in V2.

- `maintainVisibleContentPosition`: https://github.com/Shopify/flash-list/issues/547#issuecomment-2851353607
- `experimentalScrollPositionManagement`: https://github.com/Shopify/flash-list/pull/824#issuecomment-2937331570

I think we should be able to continue this migration then, right? Alternatively, i don't know if we have investigated migrating to Legend List instead. @mountiny do you have any context on that?

cc @hannojg @janicduplessis @perunt

**Comment by chrispader** [comment]
> Right now, the combination of maintainVisibleContentPosition. startRenderingFromBottom + reversed data array should be used.

@VickyStash do you think your changes will conflict with my work in https://github.com/Expensify/App/pull/51366? There i'm working on some custom initial rendering logic, to allow linking to a specific report action using `maintainVisibleContentPosition`. `maintainVisibleContentPosition` should now work in V2 of FlashList, right?

Do you think with the reversed data approach will work with the changes in the other PR?

> Deep linking to the specific message <- the custom logic that we have should be adjusted

What exactly are you adjusting here? This seems to also be tightly coupled to what i'm doing there

**Comment by VickyStash** [comment]
> @VickyStash do you think your changes will conflict with my work in https://github.com/Expensify/App/pull/51366? There i'm working on some custom initial rendering logic, to allow linking to a specific report action using `maintainVisibleContentPosition`. `maintainVisibleContentPosition` should now work in V2 of FlashList, right?

It's highly possible there could be conflicts.
`maintainVisibleContentPosition` is supported in V2 of FlashList, but it doesn't allow to set `minIndexForVisible` directly (see the [props](https://shopify.github.io/flash-list/docs/usage#maintainvisiblecontentposition)). On the `FlashList` side, they always [force the maintainVisibleContentPosition to 0](https://github.com/Shopify/flash-list/blob/921847d35cca043f438a441e9348bad358ffc0fd/src/recyclerview/RecyclerView.tsx#L477-L478).

> What exactly are you adjusting here? This seems to also be tightly coupled to what i'm doing there

It just doesn't work in the expected way for now with the FlashList, if I deep link to the message, then scrolling doesn't work in the way it should.

**Comment by VickyStash** [comment]
**Updates:**

Right now, I'm mostly focused on active testing and fixing anything I find.
As for now, the most problematic things to address:
- performance regression when I switch between two chats with a lot of expenses (with 50+ expenses)
- keyboard behaviour on mobile platforms
- bug with scrolling after deeplinking to the message -> scrolling to trigger paging -> scrolling to the bottom of the list

The plan is to try to resolve these issues and then, if nothing critical, prepare the Draft PR to get more feedback/testing.

**Comment by VickyStash** [comment]
**Note:**

> Overall, it seems like our [PureReportActionItem](https://github.com/Expensify/App/blob/main/src/pages/home/report/PureReportActionItem.tsx) component is too complex for the recycling mechanism.

It turned out not to be the root cause!
I've noticed one more thing: when I switch from Chat 1 to Chat 2, messages from Chat 1 are re-rendered (though they are not displayed). And then when I change screen dimensions for example on the Chat 2, Chat 1 messages re-renders are triggered.
The same happens on the main with `FlatList` as well, it's just not so noticeable. I'm checking it all.

**Comment by VickyStash** [comment]
### There are several issues as for now, that could stop us from migration:

### 1. List flickering during rendering if `maintainVisibleContentPosition.startRenderingFromBottom` is set to true (inverted list).

- **Production behaviour (FlatList)**
     
     - **Description:**  when switching between chats, messages are not flickering

        <details>
        <summary>video</summary>
    
        https://github.com/user-attachments/assets/110430ee-72b0-498d-a260-5c9ecefa16d7
    
        </details>
    
 - **FlashList behaviour**

     - **Description:**  when switching between chats, only the bottom message appears firstly, all other messages are rendered  later

        <details>
        <summary>video</summary>

        https://github.com/user-attachments/assets/ca7b9e54-b045-4bbf-b3c9-d862c3fb9340

    
        </details>

    - **Why it happens?**
    
     It seems like due to the way how rendering calculations are working  under the hood of FlashList when `maintainVisibleContentPosition.startRenderingFromBottom` is set to true.

| With `maintainVisibleContentPosition.startRenderingFromBottom` | Without `maintainVisibleContentPosition.startRenderingFromBottom` |
| ------------- | ------------- |
| https://github.com/user-attachments/assets/362827eb-6e0d-4a6f-913c-fb9dc6749775  | https://github.com/user-attachments/assets/4175831a-3530-4752-b1df-7e440a7f8779 |



### 2. Keyboard behaviour

**Production:**

The keyboard always shifts the content to the top.

<details>
<summary>video</summary>
    
https://github.com/user-attachments/assets/a18d41ba-9704-4074-9f3e-8eae46257488
    
</details>



**FlashList:**

The keyboard shifts the content to the top only if the user is at the bottom of the chat. Otherwise, the keyboard just opens and hides the currently visible messages.

<details>
<summary>video</summary>
    
https://github.com/user-attachments/assets/261b21ae-15e6-47fe-8975-2c30cc2599a7
    
</details>

I've tried to change the behaviour, but any customization made it worse.


**Reported issues:**

- https://github.com/kirillzyusko/react-native-keyboard-controller/issues/1244
- https://github.com/kirillzyusko/react-native-keyboard-controller/issues/1110
- https://github.com/Shopify/flash-list/issues/1717



### 3. Reassure test failing
I still need to define the reason of it

**Comment by mountiny** [comment]
Great job Vicky!

**Comment by mountiny** [comment]
Yeah that is not a good one, I assume we will not want to change the behaviour

**Comment by VickyStash** [comment]
> Although, I still think we should be able to make the FlashList work the same as FlatList, I dont really see a reason why it should not be possible and if there are some patches/ upstream prs required, we can try to do that

Got it, I'll continue to focus on that, mainly on [this blocker](https://github.com/Expensify/App/issues/33725#issuecomment-3824695260).

**Comment by VickyStash** [comment]
**Updates:**
- Migrated the `MoneyRequestReportPreviewContent` to FlashList ([Draft PR](https://github.com/Expensify/App/pull/81566)). Fixed ui issues with:
   - switching between images using arrows on web
   - last item display on ios
 - removed cachePolicy=none

When `FlashList` is used, there is a known issue that [Wrong images loaded when scrolling (need to wait a little for the correct images to load)](https://github.com/Shopify/flash-list/issues/749). It's even visible in the _current staging app_ version in the places `FlashList` is used. See the video (pay attention to the receipts of top elements, the wrong ones are shown after I'm back to the top cc @mountiny):

https://github.com/user-attachments/assets/e0575458-254f-4a1c-a6e1-d303e265ff3a


On ios and android we use `expo-image` lib that supports [recyclingKey](https://docs.expo.dev/versions/latest/sdk/image/#recyclingkey) prop to resolve it and I've used it.
But on the web, we use simple RN Image, so there is no built-in solution. For now, I've just added a [key prop](https://github.com/Expensify/App/pull/81566/changes#diff-d71ce45656e6f5b107343d4946055ff8e867a0f0b09ac94bed861d0b78cddb9cR38-R39) to web BaseImage, so it's not recycled, but I have an idea of how it can be adjusted.

TODOs:
- improve images recycling in the FlashList on web
- test android platform
- reduce receipt-body.svg loading size inside eReceipts
- prepare PR for the review

**Comment by mountiny** [comment]
Thanks, that looks great. Seems like expo-image is not the best for web. Or what are the reasons we did not use it there too? cc @staszekscp 

Eitherway seems like you are on  good track

**Comment by WojtekBoman** [comment]
@mountiny We currently use `Image` component from `react-native` on web to display images. I remember that when we migrated the project to `expo-image`, there were problems with changing the color of icons (it's worth mentioning that this was two years ago).

If these problems no longer occur and `expo-image` could solve the above problem, it may be worth considering using `expo-image` on web as well

**Comment by staszekscp** [comment]
It's also worth to make sure it supports all the formats we need on web vs RN Image, just to not be surprised if eg. `.heic` format has worked before, and on `expo-image` it won't

**Comment by mountiny** [comment]
> @mountiny Is it okay if I set fill color for the receipt-body.svg [right in the file](https://github.com/Expensify/App/pull/81566/changes#diff-7374eb457a92086bd5e3ba2953a4249888b948a005dc7bbc9cb37d970e723e32) at least for now? It's used in one place only and has the same fill color in both light and dark themes.

Yeah I think that if it is not dynamic, we should do that 

> @mountiny Should I include updating web images to use expo-image in the same PR? It feels like it's better to do it in a separate one to be safe, and for now, I can just use key prob to escape wrong images display.

Lets do separate and confirm the HEIC

**Comment by VickyStash** [comment]
> Could you give it a perspective on performance? I think that will be a stronger argument, let's say the user has Workspace chat with 10 reports and each has 20 expenses. What would be the difference if we limit this to 5 instead of 10?

For the context, FlashList on its own controls how many items to render, but without any customization from our side recycling of images is not looking nice to users (too active images reloading):

<details>
<summary>Video</summary>

https://github.com/user-attachments/assets/abe5a135-2a23-478a-8473-551c33dd6c93

</details>

To make it better for users, I'm applying [drawDistance](https://shopify.github.io/flash-list/docs/1.x/usage#drawdistance) prop:

<details>
<summary>Video</summary>

https://github.com/user-attachments/assets/7bc7e92e-7019-4bcf-b048-c6686477672f

</details>


So here is the comparison of the renders if 10 vs 5 expenses previewes are going to be displayed in the nested list with `drawDistance` set:

<img width="898" height="790" alt="Image" src="https://github.com/user-attachments/assets/70123903-0878-4b5e-ba57-cfef81fec213" />

**Comment by mountiny** [comment]
Asking the team about chaging the limit

**Comment by mountiny** [comment]
The caching seems like a problem on its own, could someone look into this? without caching that is a lot of work to be done with each render I am sure

**Comment by VickyStash** [comment]
- Updates regarding proposal can be found in the related [slack thread](https://expensify.slack.com/archives/C05LX9D6E07/p1771428903641059)

**Next steps:**
- update web images to use `expo-image` to support `recyclingKey`

**Comment by mountiny** [comment]
Nice progress Vicky!

**Comment by VickyStash** [comment]
**Updates:**
- Fixed the last bug in https://github.com/Expensify/App/pull/82764, asked for re-review
- Returned back to FlashList [migration PR](https://github.com/Expensify/App/pull/80140), resolved conflicts, checking everything after merging the latest main.

**Comment by VickyStash** [comment]
**Updates:**
- Answered some questions and applied updates in [Freeze PR](https://github.com/Expensify/App/pull/85443) 

**Migration [Draft PR](https://github.com/Expensify/App/pull/85114):**

Today I've been testing web, mweb, iOS, and iOS platforms.
- Added small perf improvement.
- Resolved TS issue.
- Fixed linked message opening in the middle of the huge chat when the chat is opened for the first time.

✅  web 
✅  iOS 
🟡  iOS web: found jumping issue when switching between transactions

I'll be OOO on Moday (Easter Monday), so will return to this on Tuesday

**Comment by Julesssss** [comment]
- https://github.com/Expensify/App/pull/85443 merged, keeping an eye on regressions 🤞

**Comment by mountiny** [comment]
Nice progress!

**Comment by VickyStash** [comment]
The PR is ready for review: https://github.com/Expensify/App/pull/85114

**Comment by VickyStash** [comment]
**Updates:**
- While FlashList PR is under review I've returned to this investigation:

> Experimented with maintainVisibleContentPosition to simplify opening transaction thread at top logic, but it turned out to be not so straightforward. So I've kept it as is for now, maybe we can do it as a follow up

It works +- okay on web and android web, but has issues on other platforms, I'm experimenting to see if it's possible.

**Comment by VickyStash** [comment]
**Updates:**
- Added some perfomance measurements info: https://github.com/Expensify/App/pull/85114#issuecomment-4251058456
- Experimanted more with `maintainVisibleContentPosition` to simplify opening transaction thread at top. After some adjustments on the FlashList side, the behavior improved, but there are still jumps on the initial data load inside the transaction thread. Meanwhile, @sosek108 is also doing some investigations in the same area and is going to suggest some perf improvements.

**Comment by mitarachim** [comment]
Deploy Blocker https://github.com/Expensify/App/issues/88054 was identified to be related to the PR https://github.com/Expensify/App/pull/85114 that attempted to fix this issue.

**Comment by jponikarchuk** [comment]
Deploy Blocker #88067 was identified to be related to the PR #85114 that attempted to fix this issue.

**Comment by nlemma** [comment]
Deploy Blocker https://github.com/Expensify/App/issues/88078 was identified to be related to the PR https://github.com/Expensify/App/pull/85114 that attempted to fix this issue.

**Comment by VickyStash** [comment]
Focused on deploy blockers: https://expensify.slack.com/archives/C08CZDJFJ77/p1776337234248689

**Comment by VickyStash** [comment]
Migration PR was deployed to production. List of all unsolved follow ups:
[[FlashList] Chat - Blank space appears when sending a message while scrolling up](https://github.com/Expensify/App/issues/88067)
[[FlashList] Android - Chat - Missing scroller in chat report](https://github.com/Expensify/App/issues/88054)
[[FlashList] Trip - Missing gap between these two rows on a trip card](https://github.com/Expensify/App/issues/88107) (PR exists)
[Chat Does Not Auto-Scroll or Show New Message Indicator for Incoming Messages](https://github.com/Expensify/App/issues/88238)

**Comment by sosek108** [comment]
Hi, @VickyStash is on sick leave and I'll be replacing her. If any new issue occurs, please bump me

**Comment by VickyStash** [comment]
I'm OOO till the 6th of May, if any action is needed, please reach out to @sosek108 🙏

**Comment by shubham1206agra** [comment]
In this issue, I reviewed 3 PRs
- https://github.com/Expensify/App/pull/81566
- https://github.com/Expensify/App/pull/83325
- https://github.com/Expensify/App/pull/85541

@mountiny Please recommend the payout here. Thanks

**Comment by mountiny** [comment]
I think we can go for $1000 as these 3 prs were complex

**Comment by shubham1206agra** [comment]
@mallenexpensify Can you please give a payment summary for me as per https://github.com/Expensify/App/issues/33725#issuecomment-4327703397?

**Comment by shubham1206agra** [comment]
Waiting for @mallenexpensify

**Comment by mallenexpensify** [comment]
### Payment Summary
Contributor+: @shubham1206agra due $1000 via NewDot
Contributor+: @situchan due $500 via NewDot. 

@mountiny what's @situchan due? they didn't review the same PRs as @shubham1206agra  (and there's 300+ comments on this issue). Thx

**Comment by situchan** [comment]
I also reviewed 2 complex PRs:
- https://github.com/Expensify/App/pull/82764
- https://github.com/Expensify/App/pull/85114

Follow-ups:
https://github.com/Expensify/App/pull/85443
https://github.com/Expensify/App/pull/88066
https://github.com/Expensify/App/pull/88179
https://github.com/Expensify/App/pull/88321
https://github.com/Expensify/App/pull/88529

**Comment by JmillsExpensify** [comment]
$1,000 approved for @shubham1206agra

**Comment by mountiny** [comment]
$500 to @situchan as well @mallenexpensify

**Comment by mallenexpensify** [comment]
Thanks @mountiny .  @situchan paid, summary updated above.

**Comment by laurenreidexpensify** [comment]
$500 approved for @situchan

**Comment by VickyStash** [comment]
As a follow up I've started to turn patches we have created into FlashList PRs:
https://github.com/Shopify/flash-list/pull/2096
https://github.com/Shopify/flash-list/pull/2300
I'm going to raise small PR to reference correct links instead of just TBD:
ex:
https://github.com/Expensify/App/blob/64fe1684d6bc39432c7328444801733a09ea85d6/patches/%40shopify/flash-list/details.md?plain=1#L9
